lib/jxl/base/status.h

```diff
@@ -321,6 +324,84 @@ inline JXL_FORMAT(2, 3) Status
   return status;
 }
 
+template <typename T>
+class JXL_MUST_USE_RESULT StatusOr {
+  union Storage {
+    char dummy_;
+    T data_;
+  } storage_;
+
+  static_assert(!std::is_convertible<StatusCode, T>::value &&
+                    !std::is_convertible<T, StatusCode>::value,
+                "You cannot make a StatusOr with a type convertible from or to "
+                "StatusCode");
+  static_assert(std::is_move_constructible<T>::value &&
+                    std::is_move_assignable<T>::value,
+                "T must be move constructible and move assignable");
+
+  StatusCode code_;
+
+ public:
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  StatusOr(StatusCode code) : code_(code) {
+    JXL_ASSERT(code_ != StatusCode::kOk);
+  }
+
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  StatusOr(Status status) : StatusOr(status.code()) {}
+
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  StatusOr(T&& value) : code_(StatusCode::kOk) {
+    new (&storage_.data_) T(std::move(value));
+  }
+
+  StatusOr(StatusOr&& other) noexcept {
+    if (other.ok()) {
+      new (&storage_.data_) T(std::move(other.storage_.data_));
+    }
+    code_ = other.code_;
+  }
+
+  StatusOr operator=(StatusOr&& other) noexcept {
+    if (ok() && other.ok()) {
```

Comment:
Maybe if this == &other then do nothing? Otherwise you require T::operator=(T&&) to handle aliasing itself.